errors/errors.go

```diff
@@ -0,0 +1,91 @@
+// Copyright 2019 The Go Cloud Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package errors provides an error type for Go Cloud APIs.
+package errors
+
+import "fmt"
+
+// An ErrorCode describes the error's category. Programs should act upon an error's
+// code, not its message.
+type ErrorCode int
+
+const (
+	// Returned by the Code function on a nil error. It is not a valid
+	// code for an error.
+	OK ErrorCode = 0
+
+	// The error could not be categorized.
+	Unknown = 1
+
+	// The resource was not found.
+	NotFound = 2
+
+	// The resource exists, but it should not.
+	AlreadyExists = 3
+
+	// A value given to a Go Cloud API is incorrect.
+	InvalidArgument = 4
+
+	// Something unexpected happened. Internal errors always indicate
+	// bugs in Go Cloud (or possibly the underlying provider).
+	Internal = 5
+)
+
+// To get stringer: go get golang.org/x/tools/cmd/stringer
+
+//go:generate stringer -type=ErrorCode
+
+// An Error described a Go Cloud error.
+type Error struct {
+	Code ErrorCode
+	msg  string
+	err  error
+}
+
+func (e *Error) Error() string {
+	return fmt.Sprintf("%v: %s", e.Code, e.msg)
+}
+
+// Unwrap returns the error underlying the receiver.
+func (e *Error) Unwrap() error {
+	return e.err
+}
+
+// Code returns the ErrorCode of err if it is an *Error.
+// It returns Unknown if err is a non-nil error of a different type.
+// If err is nil, it returns the special code OK.
+func Code(err error) ErrorCode {
+	if err == nil {
+		return OK
+	}
+	if e, ok := err.(*Error); ok {
+		return e.Code
+	}
+	return Unknown
+}
+
+// New returns a new error with the given code, underlying error and message.
+func New(c ErrorCode, err error, msg string) *Error {
+	return &Error{
+		Code: c,
+		msg:  msg,
```

Comment:
Do we need the additional msg? Based on the usage you showed in the other PR, it looks like an API would be sufficient? Otherwise people are mostly going to copy err.Error() as the message.

Something like:

```
func New(c ErrorCode, api string, err error) *Error {...}


// no Newf

func (e *Error) Error() string {
  return fmt.Sprintf("%s: %s - %v", e.api, e.Code, e.err)
}
```

Reply:
I think we'll want to replace a lot of fmt.Errorfs in the concrete implementation with Newfs. E.g. from blob.go:

current:
```
 return nil, fmt.Errorf("blob.NewWriter: duplicate case-insensitive metadata key %q", lowerK)
```
proposed:
```
  return nil, gcerr.Newf(gcerr.InvalidArgument, nil, blob.NewWriter: duplicate case-insensitive metadata key %q", lowerK)
```

We could also consider adding more info to driver errors:

current:
```
    if w.w, err = w.b.NewTypedWriter(w.ctx, w.key, ct, w.opts); err != nil {
        return 0, wrapError(w.b, err)
    }
```
proposed:
```
    if w.w, err = w.b.NewTypedWriter(w.ctx, w.key, ct, w.opts); err != nil {
        return 0, gcerr.Newf(w.b.ErrorCode(err), err, "writing key %q: %v", w.key, err)
    }
```
In the last case we ultimately won't need to have the "%v" with `err`, because something will format the underlying error for us. I'm not quite sure how that part is going to work at the moment, though.

Reply:
Your last paragraph is what I'm wondering about. The current code you have doesn't print the underlying error's message at all, so the caller will always want to provide a message that includes the underlying error's `Error()`, or it won't have a message at all. E.g., `gcerr.New(w.b.ErrorCode(err), err, "")` will not work. Maybe `*Error` could check is `msg` is empty, and if so, default to `err.Error`?

I also do like having the `blob: ` (or whatever) prefix for all of our errors; it's not obvious how to get that consistently if every error is formatted arbitrarily.

Maybe a separate "prefix" argument, and if "msg" is empty default to using `err.Error()`?

Reply:
Okay, I finally figured out how formatting is going to work with the new error design, and made the necessary changes. 

In the new design, `%v` prints a one-line message, and `%+v` prints details over several lines. The details will include a line like `gocloud.dev/blob.NewWriter`, so there won't be any need for a `blob:` prefix.

So say you wrap a driver error with message "foo" with a code and  message: 
`New(NotFound, err, "something")`
Output with %v:
```
NotFound: something: foo
```
Output with %+v:
```
NotFound: something
  gocloud.dev/blob.NewWriter
  path/to/file/blob.go:197
  - foo
```

If you pass `""` for the message, the output is the same except `something` won't appear.

Reply:
No, `%+v`.

Would you like to include a short API name like `blob` even for `%v`?

Reply:
It would be nice, but I guess it's not critical.